### PR TITLE
Avoid transits through `Row` for specialized types

### PR DESCRIPTION
**Pull request by vmarcos**
This PR avoids transits through `Row` for specialized types in join closures and during peek completion. This is achieved by employing `into_datum_iter` instead of `into_row` in these functions. To deal with lifetimes, the allocations of `RowArena` instances are moved earlier than obtaining key-value iterators in local scopes.

This PR builds on the design introduced by PR #21968.

### Motivation

  * This PR adds a known-desirable feature. Advances https://github.com/MaterializeInc/materialize/issues/22522

### Tips for reviewer

This PR makes progress on avoiding row encoding/decoding when unnecessary for specialized types, but is not comprehensive in getting rid of `borrow_with*` calls. These are left to future work.

One item for discussion: I did not change the definition of `IntoRowByTypes` to actually get rid of `into_row` as this API may be convenient in the future (e.g., when we are not applying an MFP in an operator). Just let me know if we'd better just get rid of it.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by vmarcos** [comment]
[A slow SLT run](https://buildkite.com/materialize/sql-logic-tests/builds/5720) look fine for this PR.

File: src/compute/src/compute_state.rs [hunk 1]
```diff
@@ -773,25 +773,23 @@ impl PendingPeek {
...
-                // if all we are doing here is returning a naturally limited number of results.
-                let key = cursor.key(&storage).into_row(&mut key_buf, key_types);
-                let row = cursor.val(&storage).into_row(&mut val_buf, val_types);
+
+                let key = cursor.key(&storage).into_datum_iter(key_types);
+                let row = cursor.val(&storage).into_datum_iter(val_types);
+
                 // TODO: We could unpack into a re-used allocation, except
                 // for the arena above (the allocation would not be allowed
                 // to outlive the arena above, from which it might borrow).
```
Comment (antiguru) [inline src/compute/src/compute_state.rs:789]: I think this comment is stale now, no?

**Review by antiguru — APPROVED** [review]
Looks good! I think removing the remaining `borrow_with*` calls is mostly mechanical, so fine to do it later.

File: src/compute/src/compute_state.rs (hunk 1 above)
Comment (vmarcos) [reply src/compute/src/compute_state.rs]: Yep, you're right, I overlooked the "except" part before. `TODO` is gone now.

**Comment by vmarcos** [comment]
TFTR!

**Pull request merged**